File: PaleoRose/Classes/Graphics/XRGraphicLine.m
//
//  XRGraphicLine.m
//  XRose
//
//  Created by Tom Moore on Mon Jan 26 2004.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "XRGraphicLine.h"
#import "XRGeometryController.h"

static NSString * const KVOKeyTickType = @"tickType";
static NSString * const KVOKeyShowTick = @"showTick";
static NSString * const KVOKeyFont = @"font";
static NSString * const KVOKeySpokeAngle = @"spokeAngle";
static NSString * const KVOKeySpokePointOnly = @"spokePointOnly";
static NSString * const KVOKeySpokeNumberOrder = @"spokeNumberOrder";
static NSString * const KVOKeySpokeNumberCompassPoint = @"spokeNumberCompassPoint";

@interface XRGraphicLine()

@property (readwrite) float relativePercent;

-(void)setLineLabel;
-(void)calculateLabelTransform;
-(void)appendParallelTransform;
-(void)appendHorizontalTransform;

@end

@implementation XRGraphicLine
-(instancetype)initWithController:(id<GraphicGeometrySource>)aController {
    if (!(self = [super initWithController:aController])) return nil;
    if(self)
    {
        self.tickType = XRGraphicLineTickTypeNone;
        _relativePercent = 1.0;
        _spokeNumberAlign = XRGraphicLineNumberAlignHorizontal;
        _spokeNumberCompassPoint = XRGraphicLineNumberPoints;
        _spokeNumberOrder = XRGraphicLineNumberingOrderQuad;
        _showLabel = YES;
        _spokePointOnly = NO;
        self.font = [NSFont fontWithName:@"Arial-Black" size:12];
        [self registerForKVO];
        [self setLineLabel];
        [self calculateGeometry];
    }
    return self;
}

-(void)registerForKVO {
    NSArray *keys = @[KVOKeyTickType, KVOKeyShowTick, KVOKeyFont, KVOKeySpokeAngle, KVOKeySpokePointOnly, KVOKeySpokeNumberOrder, KVOKeySpokeNumberCompassPoint];
    [keys enumerateObjectsUsingBlock:^(NSString * _Nonnull keyPath, NSUInteger idx, BOOL * _Nonnull stop) {
        [self addObserver:self
               forKeyPath:keyPath
                  options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld)
                  context:NULL];
    }];
}

#pragma mark - Geometry Calculation

- (void)calculateGeometry {
    self.drawingPath = [[NSBezierPath alloc] init];
    
    // Draw the inner point of the line
    [self drawLineFromCenterToRadius:[self radiusForInnerPoint]];
    
    // Draw the outer point of the line with appropriate tick adjustment
    [self drawLineToOuterPoint];
    
    [self calculateLabelTransform];
}

- (void)drawLineFromCenterToRadius:(CGFloat)radius {
    NSPoint startPoint = [self pointAtRadius:radius angle:self.spokeAngle];
    [self.drawingPath moveToPoint:startPoint];
}

- (void)drawLineToOuterPoint {
    CGFloat radius = [self calculateOuterRadius];
    NSPoint endPoint = [self pointAtRadius:radius angle:self.spokeAngle];
    [self.drawingPath lineToPoint:endPoint];
}

- (CGFloat)calculateOuterRadius {
    if (self.tickType == XRGraphicLineTickTypeNone || !_showTick) {
        return [self.geometryController radiusOfRelativePercent:_relativePercent];
    } else if (self.tickType == XRGraphicLineTickTypeMinor) {
        return [self.geometryController unrestrictedRadiusOfRelativePercent:(_relativePercent + 0.05)];
    } else {
        return [self.geometryController unrestrictedRadiusOfRelativePercent:(_relativePercent + 0.1)];
    }
}

- (CGFloat)radiusForInnerPoint {
    return [self.geometryController radiusOfRelativePercent:0.0];
}

- (NSPoint)pointAtRadius:(CGFloat)radius angle:(CGFloat)angle {
    NSPoint point = NSMakePoint(0.0, radius);
    return [self.geometryController rotationOfPoint:point byAngle:angle];
}

#pragma mark - Label Handling

- (void)setLineLabel {
    // Reset showLabel flag
    _showLabel = YES;
    
    // Handle spoke point only mode
    if (_spokePointOnly) {
        // In point-only mode, show N/S/E/W for cardinal directions when in compass point mode
        if (_spokeNumberCompassPoint == XRGraphicLineNumberPoints) {
            // For compass point mode, only show N/S/E/W for exact cardinal directions
            if (self.spokeAngle == 0.0 || self.spokeAngle == 90.0 || 
                self.spokeAngle == 180.0 || self.spokeAngle == 270.0 || 
                self.spokeAngle == 360.0) {
                [self setCompassPointLabel];
                [self calculateLabelTransform];
                return;
            }
        } 
        // For non-compass point mode, show the angle value for cardinal directions
        else if (self.spokeAngle == 0.0 || self.spokeAngle == 90.0 || 
                self.spokeAngle == 180.0 || self.spokeAngle == 270.0 || 
                self.spokeAngle == 360.0) {
            [self setLabelForAngle:self.spokeAngle];
            [self calculateLabelTransform];
            return;
        }
        
        // For all other cases in point-only mode, show empty string
        _showLabel = NO;
        _lineLabel = [[NSMutableAttributedString alloc] initWithString:@""];
        [self calculateLabelTransform];
        return;
    }
    // Handle compass points (N/S/E/W) when in compass point mode
    else if (_spokeNumberCompassPoint == XRGraphicLineNumberPoints && 
             (self.spokeAngle == 0.0 || self.spokeAngle == 90.0 || 
              self.spokeAngle == 180.0 || self.spokeAngle == 270.0 || 
              self.spokeAngle == 360.0)) {
        [self setCompassPointLabel];
    }
    // Handle quadrant-based numbering
    else if (_spokeNumberOrder == XRGraphicLineNumberingOrderQuad) {
        [self setQuadrantBasedLabel];
    }
    // Handle 360-degree numbering
    else if (_spokeNumberOrder == XRGraphicLineNumberingOrder360) {
        [self setLabelForAngle:self.spokeAngle];
    }
    // Default case - show the angle as is
    else {
        [self setLabelForAngle:self.spokeAngle];
    }
    
    [self calculateLabelTransform];
}

- (void)setCompassPointLabel {
    NSString *compassPoint = @"N"; // Default
    
    if (self.spokeAngle == 0.0 || self.spokeAngle == 360.0) {
        compassPoint = @"N";
    } else if (self.spokeAngle == 90.0) {
        compassPoint = @"E";
    } else if (self.spokeAngle == 180.0) {
        compassPoint = @"S";
    } else if (self.spokeAngle == 270.0) {
        compassPoint = @"W";
    }
    
    _lineLabel = [[NSMutableAttributedString alloc] initWithString:compassPoint];
}

- (void)setQuadrantBasedLabel {
    double workAngle;
    
    if (self.spokeAngle <= 90.0) {
        workAngle = self.spokeAngle;
    }
    else if (self.spokeAngle <= 180.0) {
        workAngle = 180.0 - self.spokeAngle;
    }
    else if (self.spokeAngle <= 270.0) {
        workAngle = self.spokeAngle - 180.0;
    }
    else {
        workAngle = 360.0 - self.spokeAngle;
    }
    
    [self setLabelForAngle:workAngle];
}

- (void)setLabelForAngle:(double)angle {
    // Match the original behavior for integer vs. floating point display
    if (angle == floor(angle)) {
        _lineLabel = [[NSMutableAttributedString alloc] initWithString:[NSString stringWithFormat:@"%d", (int)angle]];
    } else {
        _lineLabel = [[NSMutableAttributedString alloc] initWithString:[NSString stringWithFormat:@"%3.1f", angle]];
    }
}


-(void)drawRect:(NSRect)aRect
{
    if(NSIntersectsRect(aRect,[self.drawingPath bounds])) {
        [NSGraphicsContext saveGraphicsState];

        [self.strokeColor set];
        //NSLog(@"width %f",[self.drawingPath lineWidth]);
        [self.drawingPath stroke];
        if(self.drawsFill) {
            [self.fillColor set];
            [self.drawingPath fill];
        }
        if(_showLabel) {

            [_labelTransform concat];
            [_lineLabel drawAtPoint:NSMakePoint(0.0,0.0)];
        }
        [NSGraphicsContext restoreGraphicsState];
        [self setNeedsDisplay:NO];
    }
}

-(void)calculateLabelTransform {
    NSRange labelRange;
    labelRange.location = 0;
    labelRange.length = [_lineLabel length];
    [_lineLabel setAttributes:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:self.font,self.strokeColor,nil]
                                                          forKeys:[NSArray arrayWithObjects:NSFontAttributeName,NSForegroundColorAttributeName,nil]] range:labelRange];
    _labelTransform = [NSAffineTransform transform];
    if(_spokeNumberAlign == XRGraphicLineNumberAlignHorizontal)
        [self appendHorizontalTransform];
    else
        [self appendParallelTransform];
}

-(void)appendHorizontalTransform {
    NSSize theSize = [_lineLabel size];
    float displacement = [self.geometryController unrestrictedRadiusOfRelativePercent:(_relativePercent + .2)];
    float rotationAngle;
    //step 1. shift to the center point
    [_labelTransform translateXBy:(-0.5*theSize.width) yBy:(-0.5*theSize.height)];
    //step 2. rotation in the inverse direction
    rotationAngle = self.spokeAngle - 90;

    [_labelTransform rotateByDegrees:(-1*rotationAngle)];

    //step 3. shift out
    [_labelTransform translateXBy:displacement yBy:0.0];

    //step 4. final rotation

    [_labelTransform rotateByDegrees:rotationAngle];
}

-(void)appendParallelTransform {

    NSSize theSize = [_lineLabel size];

    float displacement = [self.geometryController unrestrictedRadiusOfRelativePercent:(_relativePercent + 0.1)];
    float rotationAngle = 90-self.spokeAngle;
    //[_labelTransform translateXBy:(-0.5*theSize.width) yBy:(-0.5*theSize.height)];
    //transform 1.  Shift the text down by half the height.
    if((double)self.spokeAngle == 0.0)
    {
        [_labelTransform translateXBy:(-0.5*theSize.width) yBy:displacement];
    }
    else if((double)self.spokeAngle == 180.0)
        [_labelTransform translateXBy:(-0.5*theSize.width) yBy:(displacement+theSize.height)*-1];
    else
    {
        if(self.spokeAngle > 180.0)
        {
            [_labelTransform rotateByDegrees:rotationAngle-180];
            [_labelTransform translateXBy:-1 *(displacement+theSize.width) yBy:0.0];
            [_labelTransform translateXBy:0.0 yBy:(-0.5*theSize.height)];
        }
        else
        {

            [_labelTransform rotateByDegrees:rotationAngle];
            [_labelTransform translateXBy:displacement yBy:0.0];
            [_labelTransform translateXBy:0.0 yBy:(-0.5*theSize.height)];
        }
    }
}

-(NSDictionary *)graphicSettings {
    NSMutableDictionary *parentDict = [NSMutableDictionary dictionaryWithDictionary:[super graphicSettings]];
    NSDictionary *classDict = @{
        XRGraphicKeyGraphicType            : GraphicTypeLine,
        XRGraphicKeyRelativePercent        : [self stringFromFloat:  _relativePercent],
        XRGraphicKeyAngleSetting           : [self stringFromFloat: self.spokeAngle],
        XRGraphicKeyTickType               : [self stringFromInt: _tickType],
        XRGraphicKeyShowTick               : [self stringFromBool:_showTick],
        XRGraphicKeySpokeNumberAlignment   : [self stringFromInt: _spokeNumberAlign],
        XRGraphicKeySpokeNumberCompassPoint: [self stringFromInt: _spokeNumberCompassPoint],
        XRGraphicKeySpokeNumberOrder       : [self stringFromInt: _spokeNumberOrder],
        XRGraphicKeyShowLabel              : [self stringFromBool:_showLabel],
        XRGraphicKeyLineLabel              : _lineLabel ?: @"",
        XRGraphicKeyCurrentFont            : self.font ?: [NSFont systemFontOfSize:[NSFont systemFontSize]]
    };
    [parentDict addEntriesFromDictionary:classDict];
    return parentDict;
}


-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    [super observeValueForKeyPath:keyPath
                         ofObject:object
                           change:change
                          context:context];
    if ([keyPath isEqualToString: KVOKeyTickType] || [keyPath isEqualToString: KVOKeyShowTick] || [keyPath isEqualToString: KVOKeyFont]) {
        [self calculateGeometry];
    } else if ([keyPath isEqualToString: KVOKeySpokeAngle] || [keyPath isEqualToString: KVOKeySpokePointOnly] ||
               [keyPath isEqualToString: KVOKeySpokeNumberCompassPoint] || [keyPath isEqualToString: KVOKeySpokeNumberOrder]) {
        [self setLineLabel];
        [self calculateGeometry];
    }
}

@end

File: PaleoRose/Classes/Graphics/XRGraphicLine.swift
//
//  XRGraphicLine.swift
//  PaleoRose
//
//  Created by Migration Assistant on 2025-08-28.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import AppKit

// MARK: - Swift Enums with Objective-C compatibility

@objc public enum XRGraphicLineTickType: Int32 {
    case none = 0
    case minor = 1
    case major = 2
}

@objc public enum XRGraphicLineNumberAlign: Int32 {
    case horizontal = 0
    case angle = 1
}

@objc public enum XRGraphicLineNumberCompassPoint: Int32 {
    case numbersOnly = 0
    case points = 1
}

@objc public enum XRGraphicLineNumberingOrder: Int32 {
    case order360 = 0
    case quad = 1
}

// MARK: - Main Class

@objc class XRGraphicLine: XRGraphic {

    // MARK: - Properties (Swift enum-based)

    // Private enum storage
    private var tickTypeEnum: XRGraphicLineTickType = .none
    private var spokeNumberAlignEnum: XRGraphicLineNumberAlign = .horizontal
    private var spokeNumberCompassPointEnum: XRGraphicLineNumberCompassPoint = .points
    private var spokeNumberOrderEnum: XRGraphicLineNumberingOrder = .quad

    // Objective-C compatible Int32 computed properties
    @objc dynamic var tickType: Int32 {
        get { tickTypeEnum.rawValue }
        set {
            tickTypeEnum = XRGraphicLineTickType(rawValue: newValue) ?? .none
            calculateGeometry()
        }
    }

    @objc dynamic var spokeNumberAlign: Int32 {
        get { spokeNumberAlignEnum.rawValue }
        set {
            spokeNumberAlignEnum = XRGraphicLineNumberAlign(rawValue: newValue) ?? .horizontal
            setLineLabel()
            calculateGeometry()
        }
    }

    @objc dynamic var spokeNumberCompassPoint: Int32 {
        get { spokeNumberCompassPointEnum.rawValue }
        set {
            spokeNumberCompassPointEnum = XRGraphicLineNumberCompassPoint(rawValue: newValue) ?? .points
            setLineLabel()
            calculateGeometry()
        }
    }

    @objc dynamic var spokeNumberOrder: Int32 {
        get { spokeNumberOrderEnum.rawValue }
        set {
            spokeNumberOrderEnum = XRGraphicLineNumberingOrder(rawValue: newValue) ?? .quad
            setLineLabel()
            calculateGeometry()
        }
    }

    // Other properties
    @objc dynamic var showTick: Bool = false {
        didSet { calculateGeometry() }
    }

    @objc dynamic var spokePointOnly: Bool = false {
        didSet {
            setLineLabel()
            calculateGeometry()
        }
    }

    @objc dynamic var showLabel: Bool = true {
        didSet {
            setLineLabel()
            calculateGeometry()
        }
    }

    @objc dynamic var spokeAngle: Float = 0.0 {
        didSet {
            setLineLabel()
            calculateGeometry()
        }
    }

    @objc dynamic var font: NSFont? = NSFont(name: "Arial-Black", size: 12) {
        didSet {
            setLineLabel()
            calculateGeometry()
        }
    }

    private var relativePercent: Float = 1.0
    @objc public var lineLabel: NSMutableAttributedString?
    @objc public var labelTransform: NSAffineTransform?

    // MARK: - Initialization

    @objc override init(controller: GraphicGeometrySource) {
        super.init(controller: controller)
        setupInitialValues()
        setLineLabel()
        calculateGeometry()
    }

    private func setupInitialValues() {
        tickTypeEnum = .none
        showTick = false
        spokePointOnly = false
        spokeNumberAlignEnum = .horizontal
        showLabel = true
        spokeNumberCompassPointEnum = .points
        spokeNumberOrderEnum = .quad
        spokeAngle = 0.0
        font = NSFont(name: "Arial-Black", size: 12)
        relativePercent = 1.0
    }

    // MARK: - Geometry Calculation

    @objc override func calculateGeometry() {
        guard let controller = geometryController else { return }

        drawingPath = NSBezierPath()

        // Calculate line from center to outer point
        let innerRadius = CGFloat(controller.radius(ofRelativePercent: 0.0))
        let outerRadius = calculateOuterRadius()

        let innerPoint = pointAtRadius(innerRadius, angle: CGFloat(spokeAngle))
        let outerPoint = pointAtRadius(outerRadius, angle: CGFloat(spokeAngle))

        drawingPath?.move(to: innerPoint)
        drawingPath?.line(to: outerPoint)
        drawingPath?.lineWidth = CGFloat(lineWidth)

        calculateLabelTransform()
    }

    private func calculateOuterRadius() -> CGFloat {
        guard let controller = geometryController else { return 0 }

        if tickTypeEnum == .none || !showTick {
            return CGFloat(controller.radius(ofRelativePercent: Double(relativePercent)))
        } else if tickTypeEnum == .minor {
            return CGFloat(controller.unrestrictedRadius(ofRelativePercent: Double(relativePercent + 0.05)))
        } else { // major
            return CGFloat(controller.unrestrictedRadius(ofRelativePercent: Double(relativePercent + 0.1)))
        }
    }

    private func pointAtRadius(_ radius: CGFloat, angle: CGFloat) -> NSPoint {
        guard let controller = geometryController else { return NSZeroPoint }
        let point = NSMakePoint(0.0, radius)
        return controller.rotation(of: point, byAngle: Double(angle))
    }

    // MARK: - Label Handling

    private func setLineLabel() {
        if spokePointOnly {
            handleSpokePointOnlyLabel()
        } else if spokeNumberCompassPointEnum == .points, isCardinalAngle(spokeAngle) {
            setCompassPointLabel()
        } else if spokeNumberOrderEnum == .quad {
            setQuadrantBasedLabel()
        } else {
            setAngleBasedLabel()
        }
        calculateLabelTransform()
    }

    private func handleSpokePointOnlyLabel() {
        if spokeNumberCompassPointEnum == .points {
            if isCardinalAngle(spokeAngle) {
                setCompassPointLabel()
                return
            }
        } else if isCardinalAngle(spokeAngle) {
            setAngleBasedLabel()
            return
        }

        // For non-cardinal angles in spokePointOnly mode, just set empty label
        // Don't modify showLabel here to avoid recursion
        lineLabel = NSMutableAttributedString(string: "")
    }

    private func isCardinalAngle(_ angle: Float) -> Bool {
        let cardinalAngles: [Float] = [0.0, 90.0, 180.0, 270.0, 360.0]
        return cardinalAngles.contains(angle)
    }

    private func setCompassPointLabel() {
        let compassPoint = switch spokeAngle {
        case 0.0, 360.0:
            "N"
        case 90.0:
            "E"
        case 180.0:
            "S"
        case 270.0:
            "W"
        default:
            "N"
        }
        lineLabel = NSMutableAttributedString(string: compassPoint)
    }

    private func setQuadrantBasedLabel() {
        let angle = Double(spokeAngle)
        let quadrantAngle: Double = if angle <= 90.0 {
            angle
        } else if angle <= 180.0 {
            180.0 - angle
        } else if angle <= 270.0 {
            angle - 180.0
        } else {
            360.0 - angle
        }

        setLabelForAngle(quadrantAngle)
    }

    private func setAngleBasedLabel() {
        setLabelForAngle(Double(spokeAngle))
    }

    private func setLabelForAngle(_ angle: Double) {
        let labelString = if angle == floor(angle) {
            String(format: "%.0f", angle)
        } else {
            String(format: "%.1f", angle)
        }
        lineLabel = NSMutableAttributedString(string: labelString)
    }

    // MARK: - Label Transform

    private func calculateLabelTransform() {
        guard let label = lineLabel, let font else { return }

        // Apply font and color attributes
        let range = NSRange(location: 0, length: label.length)
        label.setAttributes([
            .font: font,
            .foregroundColor: strokeColor ?? NSColor.black
        ], range: range)

        labelTransform = NSAffineTransform()

        if spokeNumberAlignEnum == .horizontal {
            calculateHorizontalTransform()
        } else {
            calculateParallelTransform()
        }
    }

    private func calculateHorizontalTransform() {
        guard let controller = geometryController,
              let label = lineLabel,
              let transform = labelTransform else { return }

        let labelSize = label.size()
        let displacement = CGFloat(controller.unrestrictedRadius(ofRelativePercent: Double(relativePercent + 0.2)))
        let rotationAngle = CGFloat(spokeAngle - 90.0)

        transform.translateX(by: -0.5 * labelSize.width, yBy: -0.5 * labelSize.height)
        transform.rotate(byDegrees: -rotationAngle)
        transform.translateX(by: displacement, yBy: 0.0)
        transform.rotate(byDegrees: rotationAngle)
    }

    private func calculateParallelTransform() {
        guard let controller = geometryController,
              let label = lineLabel,
              let transform = labelTransform else { return }

        let labelSize = label.size()
        let displacement = CGFloat(controller.unrestrictedRadius(ofRelativePercent: Double(relativePercent + 0.1)))
        let rotationAngle = CGFloat(90.0 - spokeAngle)

        if spokeAngle == 0.0 {
            transform.translateX(by: -0.5 * labelSize.width, yBy: displacement)
        } else if spokeAngle == 180.0 {
            transform.translateX(by: -0.5 * labelSize.width, yBy: -(displacement + labelSize.height))
        } else {
            if spokeAngle > 180.0 {
                transform.rotate(byDegrees: rotationAngle - 180.0)
                transform.translateX(by: -(displacement + labelSize.width), yBy: 0.0)
                transform.translateX(by: 0.0, yBy: -0.5 * labelSize.height)
            } else {
                transform.rotate(byDegrees: rotationAngle)
                transform.translateX(by: displacement, yBy: 0.0)
                transform.translateX(by: 0.0, yBy: -0.5 * labelSize.height)
            }
        }
    }

    // MARK: - Drawing

    @objc override func draw(_ rect: NSRect) {
        guard let path = drawingPath else { return }

        if rect.intersects(path.bounds) {
            NSGraphicsContext.saveGraphicsState()

            strokeColor?.set()
            path.stroke()

            if drawsFill {
                fillColor?.set()
                path.fill()
            }

            if showLabel, let transform = labelTransform, let label = lineLabel {
                transform.concat()
                label.draw(at: NSZeroPoint)
            }

            NSGraphicsContext.restoreGraphicsState()
            needsDisplay = false
        }
    }

    // MARK: - Settings Export

    @objc override func graphicSettings() -> [AnyHashable: Any] {
        var settings = super.graphicSettings()

        settings["GraphicType"] = "Line"
        settings["_tickType"] = string(from: tickType)
        settings["_spokeNumberCompassPoint"] = string(from: spokeNumberCompassPoint)
        settings["_showLabel"] = string(from: showLabel)
        settings["_lineWidth"] = string(from: lineWidth)
        settings["_relativePercent"] = string(from: relativePercent)
        settings["_showTick"] = string(from: showTick)
        settings["_spokeNumberOrder"] = string(from: spokeNumberOrder)
        settings["_lineLabel"] = lineLabel?.string ?? "N"
        settings["_fillColor"] = fillColor ?? NSColor.black
        settings["_angleSetting"] = string(from: spokeAngle)
        settings["_spokeNumberAlign"] = string(from: spokeNumberAlign)
        settings["_strokeColor"] = strokeColor ?? NSColor.black
        settings["_currentFont"] = font ?? NSFont(name: "Arial-Black", size: 12)!

        return settings
    }
}
